### PR TITLE
[1.2.0] [EASY] Set TTL return [API-1081]

### LINKED ISSUE #701: Map.SetTTL does not return a boolean for indicating the existence of a an entry [API-1081]
`Map.SetTTL` should return both a boolean to indicate whether the entry for the given key exists:
* `true`: Entry exists
* `false`: Entry does not exist

Since this is public API, we should introduce a new method.

### PR DESCRIPTION
**Pull request by utku-caglayan**
Fixes #701

This work introduces "SetTTLAffected" functionality to "Map" API. We are introducing this to let users access to the boolean response of the server. We did not want to break the existing api, hence this is created as a new function.

### CODE REVIEW

File: map_it_test.go [hunk 1]
```diff
@@ -1047,7 +1047,7 @@ func TestMap_SetTTL(t *testing.T) {
 		ctx := context.Background()
 		targetValue := "value"
 		it.Must(m.Set(ctx, "key", targetValue))
-		if err := m.SetTTL(ctx, "key", 20*time.Second); err != nil {
+		if err := m.SetTTL(ctx, "key", 2*time.Second); err != nil {
```
Comment (yuce) [inline map_it_test.go:1050]: Why the change?

File: map_it_test.go [hunk 2]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
 	})
 }
 
+func TestMap_SetTTLAffected(t *testing.T) {
+	it.MapTester(t, func(t *testing.T, m *hz.Map) {
+		ctx := context.Background()
+		testcases := []struct {
+			key        string
+			isEffected bool
```
Comment (yuce) [inline map_it_test.go:1065]: Isn't `affected` more appropriate?

File: map_it_test.go [hunk 3]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
...
+				expectErr:  false,
+			},
+			{
+				key:        "setTTL on already expired key",
+				isEffected: false,
+				expectErr:  false,
+			},
+		}
+		_ = it.MustValue(m.Put(ctx, "happy path", "someValue"))
+		_ = it.MustValue(m.PutWithTTL(ctx, "setTTL on already expired key", "someValue", time.Millisecond))
```
Comment (yuce) [inline map_it_test.go:1084]: `setTTL` => `PutWithTTL`?

File: map_it_test.go [hunk 4]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
...
+		time.Sleep(time.Millisecond)
+		for _, test := range testcases {
+			effected, err := m.SetTTLAffected(ctx, test.key, time.Second)
```
Comment (yuce) [inline map_it_test.go:1087]: `effected` => `affected`?

File: map_it_test.go [hunk 5]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
 	})
 }
 
+func TestMap_SetTTLAffected(t *testing.T) {
+	it.MapTester(t, func(t *testing.T, m *hz.Map) {
+		ctx := context.Background()
+		testcases := []struct {
```
Comment (yuce) [inline map_it_test.go:1063]: How about `testCases`?

File: map_it_test.go [hunk 6]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
...
+			{
+				key:        "setTTL on already expired key",
+				isEffected: false,
+				expectErr:  false,
+			},
+		}
+		_ = it.MustValue(m.Put(ctx, "happy path", "someValue"))
+		_ = it.MustValue(m.PutWithTTL(ctx, "setTTL on already expired key", "someValue", time.Millisecond))
+		time.Sleep(time.Millisecond)
+		for _, test := range testcases {
```
Comment (yuce) [inline map_it_test.go:1086]: How about `tc` instead of `test`?

File: map_it_test.go [hunk 7]
```diff
@@ -1057,6 +1057,40 @@ func TestMap_SetTTL(t *testing.T) {
 	})
 }
 
+func TestMap_SetTTLAffected(t *testing.T) {
+	it.MapTester(t, func(t *testing.T, m *hz.Map) {
+		ctx := context.Background()
+		testcases := []struct {
+			key        string
+			isEffected bool
+			expectErr  bool
```
Comment (yuce) [inline map_it_test.go:1066]: `Err` prefixed identifiers are usually `error`s. How about renaming that to something like `failureExpected`?

**Review by yuce — CHANGES_REQUESTED** [review]
Looks good, expect a few minor things.

File: map_it_test.go (hunk 1 above)
Comment (utku-caglayan) [reply map_it_test.go:1050]: I think this takes too much time on test pipeline. Am I mistaken?

File: map_it_test.go (hunk 3 above)
Comment (utku-caglayan) [reply map_it_test.go:1084]: It is just to populate the map. Actual test case is in the for loop body.
If you mean something else can you explain?

File: map_it_test.go (hunk 7 above)
Comment (utku-caglayan) [reply map_it_test.go:1066]: Since we are asserting for an expected error, I do not think renaming it as "failure" is a good practice. Sadly I do not have a reference to defend this point. If I can find an example for this on standard library, I will share it

Comment (utku-caglayan) [reply map_it_test.go:1066]: https://github.com/golang/go/blob/master/src/crypto/x509/name_constraints_test.go#L46
This is an example from std lib.

File: map_it_test.go (hunk 1 above)
Comment (yuce) [reply map_it_test.go:1050]: 2 seconds is not enough for the TTL behaviour. That will end up with a flaky test.

File: map_it_test.go (hunk 7 above)
Comment (yuce) [reply map_it_test.go:1066]: You are not asserting for an expected error, but the existence of an error. The field in the link you mention is a string field (presumably the contents of an error message). How about `errorExpected` then?

File: map_it_test.go (hunk 3 above)
Comment (yuce) [reply map_it_test.go:1084]: OK, that key is a bit confusing. How about renaming it something else, maybe something random?

File: map_it_test.go (hunk 1 above)
Comment (utku-caglayan) [reply map_it_test.go:1050]: Got it, I was not aware of that. Asking just for curiosity, is this a documented behavior?
I would love to check if there is a document for these in the future, before making such changes.

File: map_it_test.go (hunk 7 above)
Comment (utku-caglayan) [reply map_it_test.go:1066]: That sounds good to me, thanks

File: map_it_test.go (hunk 3 above)
Comment (utku-caglayan) [reply map_it_test.go]: Thanks, fixed as suggested!

**Review by yuce — APPROVED** [review]
Thanks, looks good!

**Review by alisengul53 — APPROVED** [review]
Looks good to me
